src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ChangeLogMigrationService.kt

```diff
@@ -0,0 +1,48 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service
+
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.dto.ReferralAmendmentDetails
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.AuthUser
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Changelog
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ReferralDetails
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ChangelogRepository
+import java.time.OffsetDateTime
+import java.util.UUID
+
+@Component
+class ChangeLogMigrationService(val changelogRepository: ChangelogRepository) {
+
+  fun logChanges(referralDetails: ReferralDetails, amendValue: String, topic: AmendTopic, actor: AuthUser, reasonForChange: String): Changelog {
+    return if (topic == AmendTopic.COMPLETION_DATETIME) {
+      val oldValue = ReferralAmendmentDetails(listOf(referralDetails.completionDeadline.toString()))
+      val newValue = ReferralAmendmentDetails(listOf(amendValue))
+      processChangeLog(topic, oldValue, newValue, referralDetails.referralId, actor, reasonForChange)
+    } else {
+      val oldValue = ReferralAmendmentDetails(listOf(referralDetails.maximumEnforceableDays.toString()))
+      val newValue = ReferralAmendmentDetails(listOf(amendValue))
+      processChangeLog(AmendTopic.MAXIMUM_ENFORCEABLE_DAYS, oldValue, newValue, referralDetails.referralId, actor, reasonForChange)
```

Comment:
_unsafe branching_

Any `AmendTopic` can be passed into this function; if `topic != AmendTopic.COMPLETION_DATETIME` is true, we'll log it with `AmendTopic.MAXIMUM_ENFORCEABLE_DAYS` even if the topic isn't that.

While there's a `referralDetails` argument, it sounds reasonable to explicitly block invalid values at the point of `logChanges`, not before.

What do you think about either

- ensure that the topic one of `COMPLETION_DATETIME` or `MAXIMUM_ENFORCEABLE_DAYS`
- or detect the topic from `referralDetails`, by passing down the previous value? It would seem that would lead to better cohesion

Reply:
Sure @sldblog I got your point. The reason I did this because of both point where this method is been called, its obvious that we will get only `COMPLETION_DATETIME` or `MAXIMUM_ENFORCEABLE_DAYS`. Sure, We will get this done in a different PR